pkg/cloud/gcp/actuators/util/gcp_tags_labels.go

```diff
@@ -119,3 +162,118 @@ func GetLabelsList(userLabelsAllowed bool, client controllerclient.Client, machi
+}
+
+func (c *tagValuesClient) getTagValuesNames(ctx context.Context, tagList []machinev1.ResourceManagerTag) (map[string]string, error) {
+	tagValueList := make(map[string]string, len(tagList))
+	getTagValuesReq := &rscmgrpb.GetNamespacedTagValueRequest{}
+	for _, tag := range tagList {
+		getTagValuesReq.Name = fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		value, err := c.GetNamespacedTagValue(ctx, getTagValuesReq)
+		if err != nil {
+			var gErr *apierror.APIError
+			// google API returns StatusForbidden or StatusNotFound when the tag
+			// does not exist, since it could be because of permission issues
+			// or genuinely tag does not exist.
+			if errors.As(err, &gErr) && (gErr.HTTPCode() == http.StatusNotFound ||
+				gErr.HTTPCode() == http.StatusForbidden) {
+				klog.Errorf("does not have permission to access %s tag or tag does not exist", getTagValuesReq.Name)
+				continue
+			}
+			return nil, err
+		}
+		tagValueList[value.Parent] = value.Name
+	}
+
+	return tagValueList, nil
+}
+
+// mergeInfraProviderSpecTags merges user-defined tags in Infrastructure.Status and
+// GCPMachineProviderSpec, with precedence given to those in GCPMachineProviderSpec
+// for new or updated tags. infraTags arg will contain the merged list.
+func mergeInfraProviderSpecTags(infraTags *[]machinev1.ResourceManagerTag, providerSpecTags []machinev1.ResourceManagerTag) {
+	for _, pTag := range providerSpecTags {
+		appendTag := true
+		for i, iTag := range *infraTags {
+			if pTag.ParentID == iTag.ParentID && pTag.Key == iTag.Key && pTag.Value == iTag.Value {
+				appendTag = false
+				break
+			}
+			if pTag.ParentID == iTag.ParentID && pTag.Key == iTag.Key && pTag.Value != iTag.Value {
+				appendTag = false
+				(*infraTags)[i].Value = pTag.Value
+				break
+			}
+		}
+		if appendTag {
+			*infraTags = append(*infraTags, pTag)
+		}
+	}
+}
+
+// GetResourceManagerTags returns the merged list of user-defined tags in Infrastructure.Status
+// and GCPMachineProviderSpec to apply on the resources.
+func (t *tagManager) GetResourceManagerTags(ctx context.Context) (map[string]string, error) {
+	if !t.tagsFeatureEnabled {
+		klog.V(1).Infof("getResourceManagerTags: GCP tags featureGate is not enabled, skip adding tags")
+		return nil, nil
+	}
+
+	infra, err := GetInfrastructure(t.client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cluster infrastructure: %w", err)
+	}
+	userTags := getInfraResourceTagsList(infra.Status.PlatformStatus)
+
+	if len(userTags) == 0 && len(t.machineProviderSpecTags) == 0 {
+		klog.V(1).Infof("getResourceManagerTags: user-defined tags in Infrastructure and MachineProviderSpec is empty")
+		return nil, nil
+	}
+
+	mergeInfraProviderSpecTags(&userTags, t.machineProviderSpecTags)
+
+	if len(userTags) > 50 {
+		return nil, fmt.Errorf("maximum of 50 tags can be added to a compute instance, "+
+			"infrstructure.status.resourceTags and Machine.Spec.ProviderSpec.UserTags put "+
```

Comment:
Please use the lower camelCase versions of the field names, they should not be capitalised

Also, Infrastructure is misspelt

Reply:
suggestion incorporated.